File: 백준/Gold/12865. 평범한 배낭/평범한 배낭.java
import java.io.*;
import java.util.*;

// 구하고자 하는 것: 배낭에 넣을 수 있는 물건들의 가치 최댓값
/**
 * d[k] : 무게가 k일 때 물건들의 가치 최댓값
 * d[k] = max(d[k], d[k - w[i]] + v[i])
 * 무게가 k일 때 물건 가치가 최대가 아닐 수 있음.
 * 근데 이렇게 하면 물건들을 한 번씩 밖에 담지 못한다는 점을 만족시키지 못함.
 * 물건들을 담았는지 여부를 체크하는 방법 -> 물건마다 우리가 순서를 부여한다.
 * d[i][j] : i번까지 담을 수 있을 때, 무게가 j일 때 최대 가치
 * d[i][j]
 * 1. d[i - 1][j] 에서 i번을 담지 안는 케이스
 * 2. d[i - 1][j - w[i]]에서, i번을 담아 d[i][j]가 되는 케이스
 */

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int k = Integer.parseInt(st.nextToken());
        int[][] d = new int[n + 1][k + 1];

        int[] w = new int[n + 1];
        int[] v = new int[n + 1];

        for (int i = 1; i <= n; i++) {
            st = new StringTokenizer(br.readLine());
            w[i] = Integer.parseInt(st.nextToken());
            v[i] = Integer.parseInt(st.nextToken());
        }

        for (int i = 1; i <= n; i++) {
            for (int j = 1; j <= k; j++) {
                d[i][j] = d[i - 1][j];
                if (j - w[i] >= 0) d[i][j] = Math.max(d[i][j], d[i - 1][j - w[i]] + v[i]);
            }
        }

        int answer = 0;
        for (int i = 1; i <= n; i++) {
            for (int j = 1; j <= k; j++) {
                answer = Math.max(answer, d[i][j]);
            }
        }
        System.out.print(answer);
    }
}

File: 백준/Gold/12865. 평범한 배낭/평범한 배낭.kt
/**
 * 물건이 1개만 존재함
 * 따라서 낮은 무게에서의 선택이 미래에 영향을 줄 수 있음
 * 이전 물건의 선택이 미래에 영향을 주지 않도록 해야 함
 * d[w][v] = 해당 물건 선택하지 않은 경우 vs 선택한 경우
 * d[id][w] = d[id - 1][w] | d[id - 1][w - weights[id]] + values[id]
 */

import java.util.*

fun main() = with(System.`in`.bufferedReader()) {
    val (n, k) = readLine().split(" ").map { it.toInt() }
    val weights = IntArray(n + 1)
    val values = IntArray(n + 1)
    val d = Array(n + 1) { IntArray(k + 1) }

    repeat(n) {
        val st = StringTokenizer(readLine())
        weights[it + 1] = st.nextToken().toInt()
        values[it + 1] = st.nextToken().toInt()
    }

    for (id in 1..n) {
        for (w in 1..k) {
            d[id][w] = d[id - 1][w]
            if (w - weights[id] >= 0) d[id][w] = maxOf(d[id][w], d[id - 1][w - weights[id]] + values[id])
        }
    }

    println(d[n][k])
}